Add zha device entity

damarco opened the pull request:
## Description:

This adds a separate entity to expose device level state. The last_seen device attribute is used to derive an online/offline state.

## Checklist:
  - [x] The code change is tested and works locally.

damarco committed:
Add endpoint entity

damarco committed:
Fix lint error

Adminiuga commented:
What's the reasoning behind having this as a "ZhaEndpointEntity" and not as a "ZhaDeviceEntity"?
It is not like different endpoints on the same physical device are going to have different RSSI/LQI

as another thought, I think it would be great to have 'nwk' address included in the device_state_attributes, so it would be easier to correlate debug messages like:
```
2018-05-23 19:32:09 DEBUG (MainThread) [zigpy.zcl] [0x402e:1:0x0019] ZCL request 0x0101: [0, 4388, 0, 6, 0]
2018-05-23 19:32:09 DEBUG (MainThread) [zigpy.zcl] [0x402e:1:0x0019] No handler for cluster command 1
2018-05-23 19:32:12 DEBUG (MainThread) [zigpy.zcl] [0x402e:1:0x0019] ZCL request 0x0101: [0, 4388, 0, 6, 0]
```

damarco commented:
I initially started with a ZhaDeviceEntity, but on the other hand manufacturer, model and battery levels are cluster attributes bound to an endpoint. So (theoretically) those can differ between different endpoints. I guess having both (device and endpoint) is overkill, so I chose the endpoint entity to have a correct mapping for the mentioned attributes, though on the other hand you might have some redundant information when multiple endpoints are used. In reality most of the devices only have a single endpoint so it should not really make a difference.

damarco committed:
Add nwk address as device state attribute

dshokouhi commented:
> As for the battery status, we already get battery sensors in HA. We could include battery status in ZhaDeviceEntity, but how much benefit would it provide? I'd rather see there hardware/firmware versions there.

We actually don't have any battery status for `zha` devices yet.  There was a PR submitted a few months ago but it got closed out.  The code may make it back but it was only for Centralite devices.  I think any level of battery reporting should be done.

Adminiuga commented:
> We actually don't have any battery status for ` zha ` devices yet. There was a PR submitted a few months ago but it got closed out. The code may make it back but it was only for Centralite devices.

Yep, you're right. I forgot that PR didn't get merge. I'm running it in my private branch and so far I get 100% battery reports. Guess need to wait a bit to drain the battery, because it's coded for 2.8V max battery voltage and currently my battery measures 3V

> I think any level of battery reporting should be done.

Agreed, but personally I like the battery in a form of a sensor. Although I'm open to idea of having battery indicator as an attribute of a ZhaEndpoint/DeviceEntity. Maybe should add another binary_sensor "battery_low"...

dmulcahey commented:
@damarco Great work. I think this is going to be very helpful. Question... how come these are hidden by default? The zwave component already does something similar to this and those entities are visible by default. Shouldn't these be visible to stay consistent w/ other components that have similar functionality?

dshokouhi commented:
I agree with @dmulcahey I actually like making it similar to zwave.

  @dmulcahey any plans to add in battery sensors again or do you think they would be better fit in this entity?  I remember seeing some mention about where best to place it.

dmulcahey commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+    def hidden(self) -> bool:
+        """Hide by default."""
+        return True
+
```
```
    @property
    def force_update(self) -> bool:
        """Force update"""
        return True
```

I think adding this would make sense. Otherwise the time under the entity name in mass and the last seen attribute don't match and that gives us a slightly confusing UI experience. 

![screen shot 2018-05-27 at 8 25 48 am](https://user-images.githubusercontent.com/1335687/40586086-453ba284-618a-11e8-9ade-bff0a614ef82.png)

with this addition they will always stay in sync.

dmulcahey commented:
There is an interesting side effect in the last_seen / hass state update logic for polled entities. It creates a slightly confusing ui: 

![screen shot 2018-05-27 at 8 25 48 am](https://user-images.githubusercontent.com/1335687/40586207-34730160-618d-11e8-82f2-13f55db3d9b2.png)

The screenshot was taken at 8:25:48 am (11 seconds after the last_update time) because the device is polled for electrical measurements it is updated every ~30 seconds for a new current reading. last_seen updates accordingly but because the state of this entity doesn't "update" the time at the top isn't updated and it's a bit misleading. 

I'm not sure what the best way to handle this is. I was going to suggest setting force_update  to true on the entity but that creates the reverse of the issue for non polled entities.

damarco commented:
Thanks for all the comments.

- Visibility: I changed this to hidden by default, because I thought the information exposed by this entity will mainly be used for automations and not necessarily directly in the GUI.

- Battery levels: I have seen different ways of reporting battery levels from different vendors so I think this is something that should be handled by the quirks framework in zigpy. Also because calculating a relative battery value very much depends on the battery type and brand. The actual values could then be exposed as CustomEndpoint properties and added to this entity.

- Endpoint or device entity: I am open to changing this to a ZhaDeviceEntity, but what would be the actual name of the entity? I wanted to use the same naming convention that was used for other entities (manufacturer_model_ieee) so that entities referring to the same device could easily be identified by matching its names. As both manufacturer and model are attributes of an endpoint those might differ for devices with multiple endpoints. I have seen devices with multiple endpoints that only had a cluster zero on one endpoint for example.

- Update status: I do not really see this as an issue as long as you look at the "x minutes ago" referring to an actual state change (not a device_state_attribute change). I think for the user the important information is the online/offline status so having an indication when this changed the last time might be useful. If one really wants to know when the last message was received from the device one can refer to the last_update attribute.

dmulcahey commented:
@damarco 

> Battery levels: I have seen different ways of reporting battery levels from different vendors so I think this is something that should be handled by the quirks framework in zigpy. Also because calculating a relative battery value very much depends on the battery type and brand. The actual values could then be exposed as CustomEndpoint properties and added to this entity.

I see your point here and I agree. 

> As both manufacturer and model are attributes of an endpoint those might differ for devices with multiple endpoints. I have seen devices with multiple endpoints that only had a cluster zero on one endpoint for example.

I have yet to see a device where the manufacturer and model differ on endpoints. I think it is very unlikely that there would ever be a single device with endpoints where this information differs.

> but what would be the actual name of the entity? I wanted to use the same naming convention that was used for other entities (manufacturer_model_ieee) so that entities referring to the same device could easily be identified by matching its names.

This is why I suggested locking the creation of these to the ZHA endpoint by profile id. This allows everything else to stay the same as it is currently and you only get 1 per device. I have devices from several manufacturers: centralite, sengled, climax technologies, securifi, SmartThings, etc. Most of them have a manufacturer specific endpoint that repeats everything on the ZHA endpoint. because of this multiple entities are created with exactly the same information with this change. I can see this leading to user confusion even with these hidden by default (which I am still not convinced that they should be.) Also, every attribute on this entity is actually pulled from the device. You're using the endpoint to back up to the device to get the data. so why not try to make these represent a device since that's what it seems to be doing anyway...

> Update status: I do not really see this as an issue as long as you look at the "x minutes ago" referring to an actual state change (not a device_state_attribute change). I think for the user the important information is the online/offline status so having an indication when this changed the last time might be useful. If one really wants to know when the last message was received from the device one can refer to the last_update attribute.

I agree that if you know what you're looking at and you have some knowledge of the internals that it makes sense. Most users won't have that knowledge and I think that this has the potential to confuse casual users. 

> Visibility: I changed this to hidden by default, because I thought the information exposed by this entity will mainly be used for automations and not necessarily directly in the GUI.

I understand. My view here is that consistency leads to a better overall user experience. zwave has already created this type of entity and they are visible by default.

Adminiuga commented:
> Battery levels: I have seen different ways of reporting battery levels from different vendors so I think this is something that should be handled by the quirks framework in zigpy. Also because calculating a relative battery value very much depends on the battery type and brand. The actual values could then be exposed as CustomEndpoint properties and added to this entity.

Per ZCL specs, the "Battery Settings Attribute Set" should be writable. If we could allow users to set "BatterySize" (attr id 0x0032) and/or "BatteryRatedVoltage" (attr id 0x0034) then battery sensor could do some approximation based on the above and current voltage. 

>>As both manufacturer and model are attributes of an endpoint those might differ for devices with multiple endpoints. I have seen devices with multiple endpoints that only had a cluster zero on one endpoint for example.

>I have yet to see a device where the manufacturer and model differ on endpoints. I think it is very unlikely that there would ever be a single device with endpoints where this information differs.

+1

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+        self._device_state_attributes = {}
+        ieee = endpoint.device.ieee
+        ieeetail = ''.join(['%02x' % (o, ) for o in ieee[-4:]])
+        if manufacturer and model is not None:
```
Are you trying to check that both manufacturer and model are not none?

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+        if self._endpoint.device.last_seen is not None:
+            time_struct = time.localtime(self._endpoint.device.last_seen)
+            update_time = time.strftime("%Y-%m-%dT%H:%M:%S", time_struct)
+        self._device_state_attributes['last_update'] = update_time
```
We should not store update time.

damarco replied:
Maybe update_time or last_update is not the best name. This is essentially the time when the last message was received from the device. I thought this might be useful for verification or troubleshooting purposes.

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+                self._state = 'offline'
+            else:
+                self._state = 'online'
+        self.schedule_update_ha_state()
```
Do not call this. This is a polling entity and so this will already be called.

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+        ieee = endpoint.device.ieee
+        ieeetail = ''.join(['%02x' % (o, ) for o in ieee[-4:]])
+        if manufacturer and model is not None:
+            self.entity_id = "{}.{}_{}_{}_{}".format(
```
There is no need to create such complicated entity ids. Since it has a unique ID, a unique entity ID will be assigned once and used every time. You don't have to assign one at all and it will just work.

damarco replied:
I wanted to keep the naming consistent with the other zha entities. Otherwise the user would have no way of identifying which entities belong together.

Adminiuga replied:
For me personally, having "manufacturer" and "model" in the entity_id provides me with a better user experience correctly identifying devices, especially if you join more than one device at a time.

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -245,6 +248,16 @@ def device_removed(self, device):
                     join,
                 )
 
+            endpoint_key = "zha-{}".format(device_key)
```
We should just use `device_key` for endpoint_key. No need to add zha to it.

damarco replied:
Wouldn't this create a conflict if the same unique_id is used twice? It is already used in the `if component: ...` block above.

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+class ZhaEndpointEntity(entity.Entity):
+    """A base class for ZHA endpoints."""
+
+    _domain = DOMAIN
```
Why?

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +372,90 @@ def zdo_command(self, tsn, command_id, args):
+            )
+
+        nwk = endpoint.device.nwk
+        self._device_state_attributes['nwk'] = '0x{0:04x}'.format(nwk)
```
If you're assigning this all in the constructor, just do it all at once.

```python
self._device_state_attributes = {
    'ieee': str(…),
    # etc
}
```

Adminiuga commented:
Why not to keep "last_seen" maybe?

On Tue, May 29, 2018, 09:59 damarco <notifications@github.com> wrote:

> *@damarco* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/zha/__init__.py
> <https://github.com/home-assistant/home-assistant/pull/14579#discussion_r191434191>
> :
>
> > +        """Return the state of the entity."""
> +        return self._state
> +
> +    @property
> +    def hidden(self) -> bool:
> +        """Hide by default."""
> +        return True
> +
> +    @property
> +    def device_state_attributes(self):
> +        """Return device specific state attributes."""
> +        update_time = None
> +        if self._endpoint.device.last_seen is not None:
> +            time_struct = time.localtime(self._endpoint.device.last_seen)
> +            update_time = time.strftime("%Y-%m-%dT%H:%M:%S", time_struct)
> +        self._device_state_attributes['last_update'] = update_time
>
> Maybe update_time or last_update is not the best name. This is essentially
> the time when the last message was received from the device. I thought this
> might be useful for verification or troubleshooting purposes.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/14579#discussion_r191434191>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFjmcKhGH7C-c5iUpNZBqjwG-ES8-aMjks5t3VRAgaJpZM4UHc_O>
> .
>

damarco committed:
Change to ZhaDeviceEntity

damarco commented:
I hope all of the comments are addressed now and this is ready to merge.

balloob commented:
I wonder if we should put this PR on hold until we have official notion of devices added to Home Assistant. Discussion here: https://github.com/home-assistant/architecture/issues/36

balloob commented on homeassistant/components/zha/__init__.py:
```diff
@@ -359,6 +376,75 @@ def zdo_command(self, tsn, command_id, args):
+        """Return device specific state attributes."""
+        update_time = None
+        if self._device.last_seen is not None:
+            time_struct = time.localtime(self._device.last_seen)
```
We should not add a last seen. This will be updated too often, causing a lot of state changes.

Adminiuga replied:
This is quite useful bit of information, especially when troubleshooting issues with your zigbee network.
What if we don't include "last_seen" while a device is healthy/online, but have it in the state attributes once a devices becomes unresponsive? at least we could see when there was last communication from the device that way?

balloob replied:
That could work, however, I think that it makes more sense if someone would contribute a management panel like we have for Z-Wave.

damarco committed:
Show last_seen only if offline

damarco committed:
Remove obsolete _discover_endpoint_info()

damarco commented:
@balloob "last_seen" is now only shown if the device is offline. I would like to add support for the new device registry in a separate PR, though as far as I understood we would need to add support for config entries first.

balloob commented:
Correct.
